source/server/debug_session_properties_restricted_service.cpp

```diff
@@ -0,0 +1,244 @@
+{
+  bool supported = false;
+  auto status = get_bool_property((NISysCfgResourceProperty)0x10001000, request->product_id(), request->serial_number().c_str(), &supported);
+  response->set_supported(supported);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::IsDebugSessionEnabled(
+  ::grpc::ServerContext* context,
+  const IsDebugSessionEnabledRequest* request,
+  IsDebugSessionEnabledResponse* response)
+{
+  uint32_t enabled = false;
+  auto status = get_uint_property((NISysCfgResourceProperty)0x10002000, request->product_id(), request->serial_number().c_str(), &enabled);
+  response->set_enabled(enabled != 0);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::IsDebugSessionServerOutOfProcess(
+  ::grpc::ServerContext* context,
+  const IsDebugSessionServerOutOfProcessRequest* request,
+  IsDebugSessionServerOutOfProcessResponse* response)
+{
+  bool out_of_process = false;
+  auto status = get_bool_property((NISysCfgResourceProperty)0x10003000, request->product_id(), request->serial_number().c_str(), &out_of_process);
+  response->set_out_of_process(out_of_process);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::SetDebugSessionEnabled(
+  ::grpc::ServerContext* context,
+  const SetDebugSessionEnabledRequest* request,
+  SetDebugSessionEnabledResponse* response)
+{
+  uint32_t enabled = request->enabled() ? 1 : 0;
+  auto status = set_uint_property((NISysCfgResourceProperty)0x10002000, request->product_id(), request->serial_number().c_str(), enabled);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::SetDebugSessionServerOutOfProcess(
+  ::grpc::ServerContext* context,
+  const SetDebugSessionServerOutOfProcessRequest* request,
+  SetDebugSessionServerOutOfProcessResponse* response)
+{
+  auto status = set_bool_property((NISysCfgResourceProperty)0x10003000, request->product_id(), request->serial_number().c_str(), request->out_of_process());
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::get_bool_property(
+  NISysCfgResourceProperty property_id,
+  uint32_t product_id,
+  const char* serial_number,
+  bool* value)
+{
+  NISysCfgStatus status = NISysCfg_OK;
+  NISysCfgSessionHandle session = NULL;
+  NISysCfgFilterHandle filter = NULL;
+  NISysCfgEnumResourceHandle resources_handle = NULL;
+  NISysCfgResourceHandle resource = NULL;
+  NISysCfgBool syscfg_value = NISysCfgBoolFalse;
+
+  try {
+    auto library = get_syscfg_library_interface();
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library->CreateFilter(session, &filter))) {
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyIsDevice, NISysCfgBoolTrue);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyProductId, product_id);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertySerialNumber, serial_number);
```

Comment:
Since you say "match all" below, I suggest only including this in the filter if a non-empty serial number string was actually passed.

Reply:
Wouldn't including the empty serial number in the filter help ensure that I'm getting the simulated device I expect?  We would only pass an empty string here if that is what we received from calling EnumerateDevices.

Reply:
I suppose it would be fine.